File: Leetcode/Dynamic Programming/#120. Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        //O(n) Solution
        int n=triangle.size();
        vector<int>cpy = triangle.back();
        for(int i=n-2; i>=0; i--){
            for(int j=0; j<triangle[i].size(); j++){
                cpy[j]=triangle[i][j]+min(cpy[j],cpy[j+1]);
            }
        }
        return cpy[0];
        /* O(n^2) Solution

        vector<vector<int>>dp(n,vector<int>(n,INT_MAX));
        dp[0][0]=triangle[0][0];
        for(int i=1; i<triangle.size(); i++){
            for(int j=0; j<triangle[i].size(); j++){
                if(j==0) dp[i][j]=triangle[i][j]+dp[i-1][j];
                else if(j==triangle[i].size()) dp[i][j]=triangle[i][j]+dp[i-1][j-1];
                else dp[i][j]=triangle[i][j]+min(dp[i-1][j-1],dp[i-1][j]);
            }
        }
        return *min_element(dp[n-1].begin(),dp[n-1].end());
        */
    }
};

File: Dynamic Programming/#120. Triangle.py
''' 1D DP
f(i, j) = ith level의 j칸에서의 min path sum

Base case
f(0, 0) = triangle[0][0]

General case
f(i, j) = triangle[i][j] + min(f(i - 1, j), f(i - 1, j - 1))
'''
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        m = len(triangle)
        dp = [inf for _ in range(m + 1)]
        dp[0] = triangle[0][0]


        for i in range(1, m):
            for j in range(i, -1, -1):
                dp[j] = triangle[i][j] + min(dp[j] if j < i else inf, dp[j - 1] if j > 0 else inf)

        return min(dp)


'''
f(i, j) = ith level의 j칸에서의 min path sum

Base case
f(0, 0) = triangle[0][0]

General case
f(i, j) = triangle[i][j] + min(f(i - 1, j), f(i - 1, j - 1))
'''
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        m = len(triangle)
        dp = [[inf for _ in range(m)] for _ in range(m + 1)]

        dp[0][0] = triangle[0][0]

        for i in range(1, m):
            for j in range(i + 1):
                dp[i][j] = triangle[i][j] + min(dp[i - 1][j], dp[i - 1][j - 1])

        return min(dp[m - 1])